web-components-f902/notes/01_html-5-build-flexible-html-with-htmltemplates-using-slots-and-web-components.md.md

```diff
@@ -0,0 +1,93 @@
+# Build Flexible HTML with HTMLTemplates using Slots and Web Components
+
+[Video link](https://egghead.io/lessons/html-5-build-flexible-html-with-htmltemplates-using-slots-and-web-components)
+
+<TimeStamp start="0:29 end="0:36">
+
+Nothing renders on the screen until we create a script that adds the template to the div.
+
+</TimeStamp>
+
+<TimeStamp start="0:39" end="0:51">
+
+We append the template's content clone node inside the div.
+
+Using `true` as the parameter to the `cloneNode()` method ensures a **deep** clone of the template's content where its subtree is also copied. If we had used `false`, only the template's content would be cloned.
+
+</TimeStamp>
+
+<TimeStamp start="0:53" end="1:00">
+
+It's worth noting **deep** clones have no effect on empty elements (such as <input> and <img> elements).
+
+</TimeStamp>
+
+<TimeStamp start="1:18" end="1:26">
+
+Read more about the `<slot>` element [on this page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot).
+
+</TimeStamp>
+
+<TimeStamp start="1:44" end="1:55">
+
+Our `<script>` element is set up with our custom element as such:
+
+```jsx
+<script>
+  const template = document.querySelector('#template');
+  class CustomElement extends HTMLElement {
+    constructor() {
+      super();
+      this.attachShadow({mode: "open"});
+      this.shadowRoot
+        .appendChild(template.content.cloneNode(true));
+    }
+  }
+  customElements.define('custom-element', CustomElement);
+</script>
+<custom-element></custom-element>
+```
+
+</TimeStamp>
+
+<TimeStamp start="1:56" end="2:07">
+
+Finally, we add our `<slot>` tags inside our template as follows:
+
+```jsx
+<template id="template">
+  <div>
+    <slot name="slot1"></slot>
+  </div>
+  <form>
+    <input/>
+  </form>
+  <slot name="slot2"></slot>
+</template>
+
+```
+
+</TimeStamp>
+
+<TimeStamp start="2:26" end"2:36">
+
+After reusing our custom element, we're able to observe how powerful `<slot>` can be. Dynamic content can be achieved with this code:
+
+<custom-element>
+    <iframe slot="slot1" width="560" height="315" src="https://www.youtube.com/embed/Bv_5Zv5c-Ts" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+    <div slot="slot2">Slot 2 text inside a div</div>
+</custom-element>
```

Comment:
This is a lot of code for a note. A good rule of thumb I've found is to keep notes to about the size of a tweet.

Reply:
Maybe instead of saying "this code:" followed by the code block, you could share a small snippet of the relevant part of the iframe. like `<iframe ... {the relevant part} ...></iframe>`

Also remember, if you use a code block to put a code fence ``` above and below the block

Reply:
Completely missed the code fence on this one lol. Thanks for the catch! And definitely a lot of code, I'll reword that.